### PR TITLE
Threadmonitor

### PR DESCRIPTION
**Pull request by oalfaro2**
Added new thread monitoring class

### CODE REVIEW

**Comment by Kakon24** [comment]
Merge conflicts in observation_run.py

**Comment by Kakon24** [comment]
Misclicked close without comment so I reopened PR:

Fix merge conflicts with observation_run.py

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.idea
```
Comment (Kakon24) [inline .gitignore:144]: Use this boiler plate .gitignore then add to it if needed.
https://github.com/github/gitignore/blob/master/Python.gitignore

**Review by Kakon24 — CHANGES_REQUESTED** [review]
Remove all .pyc, .log, .png files.  Observation ticket .jsons should be local only and not added to any remote repository.  Weather statuses as well.  Remove these from the PR and update the .gitignore to add these files so it doesn't happen in the future.

File: main/controller/thread_monitor.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+class Monitor(threading.Thread):
+    def __init__(self):
+        self.runthemonitor = True
+        self.threadcrash = threading.Event()
+        self.crashed = []
+        super(Monitor, self).__init__(name='Monitor')
+
+    def monitor_run(self, threadlist):
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: this method needs to be named just "run" to overwrite the parent class's run method, otherwise the threading won't work properly.

File: main/controller/thread_monitor.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        threadlist : LIST
+            List of thread handles
+        Returns
+        -------
+        None.
+        '''
+        logging.info('Beginning thread monitoring')
+        while self.runthemonitor:
+            for th_name in threadlist:
+                if not threadlist[th_name].is_alive():
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: is threadlist actually a list, or a dictionary or something?  if it's a normal list, this kind of indexing won't work, in this case if I understand what you're doing, I think you'd just want to do th_name.is_alive() since th_name is already an item in threadlist.

Edit:  I saw later that you are actually passing in a dictionary, which is fine, but in that case I think it'd be more clear if you made the for loop as "for th_name in threadlist.keys()" or something, so it's clear that th_name is actually just a string with the name of the thread, and calling threadlist[th_name] is indexing the dictionary.  Also update the docstring to reflect this.

File: main/controller/thread_monitor.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                    if not threadlist[th_name].name in self.crashed:
+                        self.crashed.append(threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(threadlist[th_name].name))
+                    self.threadcrash.set()
+
+    def crash_return(self):
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: I don't think you really need a method for this, you could just call Monitor.crashed to get the self.crashed list

File: main/observing/observation_run.py [hunk 5]
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.monitor_run(self.th_dict)
+
+        print('---Checking Threads--- LINE 128')
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: remember to use logging statements, never normal prints for this project

File: main/observing/observation_run.py [hunk 6]
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.monitor_run(self.th_dict)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: instead of calling monitor run, you will want to call self.monitor.start(), which will call self.monitor.run() in the background on the proper thread; the way you're doing it now, it will be called on the main thread, so all progress elsewhere will be stopped indefinitely

File: main/controller/thread_monitor.py (hunk 2 above)
Comment (Michael-Reefe) [reply main/controller/thread_monitor.py]: Consequently, the run method also can't take any arguments as far as I'm aware (other than self, of course), so I would make threadlist a self attribute instead that is created via an argument an init

File: main/observing/observation_run.py [hunk 7]
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        Checks to see if self.monitor has raised a crashed thread,
+        Restarts the crashed threads if there are any
+        
+        Returns
+        -------
+        None
+        '''
+        #print('---In thread Check')
+        threadlist = self.monitor.crash_return()
+        if self.monitor.threadcrash == True:
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: threadcrash is a threading event, not a boolean, so I don't think this will work; threading events are mainly used for wait() statements to synchronize between threads, not send true/false information, but you could instead do something like "if self.monitor.threadcrash.isSet()"

File: main/observing/observation_run.py [hunk 8]
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+            #print('Threads have crashed')
+            threadlist = self.monitor.crash_return()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: why do you grab this twice?

File: main/observing/observation_run.py [hunk 9]
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+            for thname in threadlist:
+                self.threadrestart(thname)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: make sure your method calls match the method names?  here you call "threadrestart", but I see it's defined as just "restart"

File: main/observing/observation_run.py [hunk 10]
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        -----------
+        Redefines and then restarts the inputted thread
+        
+        Parmeters
+        --------
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+
+        for x in self.th_dict:
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: self.th_dict.keys()

File: main/observing/observation_run.py [hunk 11]
```diff
@@ -630,3 +673,58 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+            print('In thread restart')
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: why don't you just check if thname == each of these strings?  the whole for loop here is pointless

Comment (Michael-Reefe) [reply main/observing/observation_run.py]: better yet, you could just do self.th_dict[thname] = Camera() if thname == 'camera' else Telescope() if thname == 'telescope' else ....
and then just self.th_dict[thname].start()

File: main/observing/observation_run.py [hunk 12]
```diff
@@ -114,6 +123,13 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.monitor_run(self.th_dict)
+
+        print('---Checking Threads--- LINE 128')
+        self.threadcheck()
+
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py:133]: I'm not sure if just scattering these about everywhere in observation_run is the best approach here, though I can't immediately think of a better alternative either since the threads do need to be restarted from observation_run itself.  In any case, I don't think you need 2 in a row here.  Maybe just put one after every weather check or something.

File: main/observing/observation_run.py [hunk 13]
```diff
@@ -449,7 +488,10 @@ def take_images(self, name, num, exp_time, _filter, end_time, path, cycle_filter
             
             if self.crash_check('MaxIm_DL.exe'):
                 continue
-            
+
+            self.crash_check('RoboFocus.exe')
+            self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: you definitely don't need more than 1 threadcheck in the same take_images loop

File: observation_tickets/Kornephoros.json [hunk 14]
```diff
@@ -0,0 +1,15 @@
+{"type": "observation_ticket",
```
Comment (Michael-Reefe) [inline observation_tickets/Kornephoros.json]: don't include any of these in commits

Comment (Michael-Reefe) [reply observation_tickets/Kornephoros.json]: make sure to remove the compiled python files and other random stuff like weather data too

**Review by Michael-Reefe — CHANGES_REQUESTED** [review]

File: main/controller/thread_monitor.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
+import time
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py:1]: the .log file and __pycache__ files still should not be included in the commit, make sure you remove those too

File: main/observing/observation_run.py [hunk 16]
```diff
@@ -114,6 +123,8 @@ def everything_ok(self):
         if message:
             logging.error('Hardware connection timeout: {}'.format(message))
 
+        self.monitor.start(self.th_dict)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: I'm pretty sure start (and therefore run) can't take any arguments (not 100% sure), but you can easily fix this by just passing in the th_dict during initialization.  Add it as a parameter to threadmonitor.\_\_init\_\_ and then when you call self.monitor = Monitor(), change to self.monitor = Monitor(th_dict)

File: main/observing/observation_run.py [hunk 17]
```diff
@@ -125,6 +136,8 @@ def everything_ok(self):
                          "a possible re-open.".format(self.config_dict.min_reopen_time))
             time.sleep(self.config_dict.min_reopen_time * 60)
             if self.conditions.sun:
+                print('---Checking Threads--- LINE 142')
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: remember to change these to logging statements; also don't include the line number, as that will change frequently as you make more edits--it's already incorrect

File: main/observing/observation_run.py [hunk 18]
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        -----------
+        Redefines and then restarts the inputted thread
+        
+        Parmeters
+        --------
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+
+        for x in self.th_dict.keys():
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: why are you looping here?  this will just restart every thread whenever it's called; you don't make reference to the passed in thname at all...shouldn't the if/else statements be asking if thname == 'camera' etc. instead?

File: main/observing/observation_run.py [hunk 19]
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        ----------
+        Checks to see if self.monitor has raised a crashed thread,
+        Restarts the crashed threads if there are any
+        
+        Returns
+        -------
+        None
+        '''
+        threadlist = self.monitor.crashed()
+        if self.monitor.threadcrash.isSet():
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: I'm not sure if you really need this event at all?  if there are any threads in the threadlist then you already know they have crashed, in this case you can just do something like:  "if threadlist:" then the for loop.

File: main/controller/thread_monitor.py [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+            List of thread handles
+        Returns
+        -------
+        None.
+        '''
+        logging.info('Beginning thread monitoring')
+        while self.run_th_monitor:
+            for th_name in threadlist.keys():
+                if not threadlist[th_name].is_alive():
+                    if not threadlist[th_name].name in self.crashed:
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: you can make this an "and" statement with the if above

one more thing:  this loop will go incredibly fast and hog a lot of CPU space while it's running...you'd better add a time.sleep for 5-15 minutes or something so it's not checking 1000000 times every second

File: main/controller/thread_monitor.py [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+                        self.crashed.append(threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(threadlist[th_name].name))
+                    self.threadcrash.set()
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: mentioned in a later comment: I'm not sure if you really need this event

File: main/observing/observation_run.py [hunk 22]
```diff
@@ -630,3 +667,54 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+            elif x == 'dome':
+                self.dome = Dome()
+                self.dome.start()
+            elif x == 'flatlamp':
+                self.flatlamp = FlatLamp()
+                self.flatlamp.start()
+            elif x == 'conditions':
+                self.conditions = Conditions()
+                self.conditions.start()
+            elif x == 'guider':
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: also, I'd make sure at the end of this method that you remove the appropriate thread from the crashed threadlist, otherwise once it crashes once it will be continuously restarted

File: main/observing/observation_run.py (hunk 16 above)
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: wait, why are you starting the threadmonitor within the everything_ok method?  you should not be doing that, everything_ok is called many times throughout an observation run.  It should be started at the same time all the other threads are started, right under the gui (line 89).  [you can't restart threads more than once - you will want to start it once at the beginning of observing and keep it going for the whole run]

File: main/observing/observation_run.py [hunk 23]
```diff
@@ -61,6 +63,7 @@ def __init__(self, observation_request_list, image_directory, shutdown_toggle, c
         self.conditions = Conditions()
         self.focuser = Focuser()
         self.flatlamp = FlatLamp()
+        self.monitor = Monitor()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: this is very nitpicky, but for organizational purposes I'd put the monitor initailization after the gui (line 72), maybe even with its own comment header saying "Initializes heartbeat monitor" or something

File: main/observing/observation_run.py [hunk 24]
```diff
@@ -125,6 +136,8 @@ def everything_ok(self):
                          "a possible re-open.".format(self.config_dict.min_reopen_time))
             time.sleep(self.config_dict.min_reopen_time * 60)
             if self.conditions.sun:
+                print('---Checking Threads--- LINE 142')
+                self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: you shouldn't be doing a threadcheck here...if you want to do one at the start of everything_ok, that's fine (~ line 129), but it shouldn't be within this if statement.  If anything, you might want to put an additional one in the while loop at line 147 since we may be waiting there for potentially hours.  same deal for the while loop at line 155, I'd add one there too.

File: main/observing/observation_run.py [hunk 25]
```diff
@@ -149,6 +162,9 @@ def everything_ok(self):
             if not self.conditions.weather_alert.isSet():
                 check = True
                 self._startup_procedure(cooler=cooler)
+
+                self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: you don't want a threadcheck here, see above comment

File: main/observing/observation_run.py [hunk 26]
```diff
@@ -176,6 +192,8 @@ def _startup_procedure(self, cooler=True):
 
         """
         initial_check = self.everything_ok()
+        print('---Checking Threads--- LINE 198')
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: assuming you fix the threadchecks in everything_ok, you don't need an additional one here, as you will have just done one within everything_ok

File: main/observing/observation_run.py [hunk 27]
```diff
@@ -229,12 +247,19 @@ def check_start_time(self, ticket):
...
+
+            print('---Checking Threads--- LINE 256')
+            self.threadcheck()
+
             logging.info("It is not the start time {} of {} observation, "
                          "waiting till start time.".format(ticket.start_time.isoformat(), ticket.name))
             current_epoch_milli = time_utils.datetime_to_epoch_milli_converter(current_time)
             start_time_epoch_milli = time_utils.datetime_to_epoch_milli_converter(ticket.start_time)
-            time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
+            #time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: why did you comment this line out?  this is a very important sleep statement, it definitely should NOT be commented out

also you don't need another threadcheck at line 262

File: main/observing/observation_run.py [hunk 28]
```diff
@@ -229,12 +247,19 @@ def check_start_time(self, ticket):
         None.
         """
         current_time = datetime.datetime.now(self.tz)
+        time.sleep(10)
         if ticket.start_time > current_time:
+
+            print('---Checking Threads--- LINE 256')
+            self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: you don't need a threadcheck here...remember these large blocks of code may look long, but most of it will be executed within milliseconds, you don't need to spam threadchecks everywhere

File: main/observing/observation_run.py [hunk 29]
```diff
@@ -257,9 +282,14 @@ def observe(self):
             self.take_calibration_images(beginning=True)
         else:
             cooler = True
-
+        
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: again this is unnecessary, the _startup_procedure method calls everything_ok, which in turn will call threadcheck

File: main/observing/observation_run.py [hunk 30]
```diff
@@ -257,9 +282,14 @@ def observe(self):
...
         self.check_start_time(self.observation_request_list[0])
         initial_shutter = self._startup_procedure(cooler=cooler)
+
+        print('---Threadcheck at 295')
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: unnecessary

File: main/observing/observation_run.py [hunk 31]
```diff
@@ -321,6 +351,7 @@ def focus_target(self, ticket):
         None.
 
         """
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: unnecessary
it's this method where you might want to do something a bit more involved, because the wait() statement at line 369 will be there for quite a while as the startup focus procedure takes ~10 images...you could add a threading event when the startup_focus_procedure is finished (we might even already have one), say it's called focused_done, and make a while loop:
while not self.focus_procedures.focused_done.isSet():
     time.sleep(60)
     self.threadcheck()

File: main/observing/observation_run.py [hunk 32]
```diff
@@ -353,7 +384,9 @@ def run_ticket(self, ticket):
             The total number of images that are specified on the
             observation ticket.
         """
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: unnecessary here

File: main/observing/observation_run.py [hunk 33]
```diff
@@ -418,6 +451,7 @@ def take_images(self, name, num, exp_time, _filter, end_time, path, cycle_filter
         i = 0
         while i < num:
             logging.debug('In take_images loop')
+            self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: this is unnecessary because we already call everything_ok once every loop

File: main/observing/observation_run.py [hunk 34]
```diff
@@ -482,6 +518,7 @@ def crash_check(self, program):
             False.
 
         """
+        self.threadcheck()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: unnecessary
you might as well just remove all the calls to threadcheck() outside of the everything_ok() method and just leave it up to everything_ok to call when it's necessary, as those are already fairly spread out throughout the code.  as long as you fix the threadcheck calls within everything_ok, as previously mentioned, it should be just fine

File: resources/weather_status/weather.txt [hunk 35]
```diff
@@ -0,0 +1 @@
...color=#ff4500> 2:08a</font></p></small></font></strong></font></td>\r\n\r\n<tr>\r\n <td  bordercolor="#000000" style="border: thin solid rgb(0,0,0)"  valign=middle><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#228b22><small><p>Low Wind Chill</p></small></font></strong></font></td>\r\n <td align=left bordercolor="#000000" style="border: thin solid rgb(0,0,0)" ><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#ff4500><small><p>46.0\xb0F<font color=#228b22>&nbsp;\r\n&nbsp;at&nbsp;&nbsp;</font><font color=#ff4500> 5:06a</font></p> </small></font></strong></font></td>\r\n </tr>\r\n<tr>\r\n <td  bordercolor="#000000" style="border: thin solid rgb(0,0,0)"  valign=middle><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#228b22><small><p>High Heat Index</p></small></font></strong></font></td>\r\n <td align=left bordercolor="#000000" style="border: thin solid rgb(0,0,0)" ><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#ff4500><small><p>58.0\xb0F<font color=#228b22>&nbsp;\r\n&nbsp;at&nbsp;&nbsp;</font><font color=#ff4500> 1:13p</font></p> </small></font></strong></font></td>\r\n </tr>\r\n\r\n </table>\r\n\r\n</BODY>\r\n</HTML>\r\n'
```
Comment (Michael-Reefe) [inline resources/weather_status/weather.txt]: this should also not be included in the commit

**Review by Michael-Reefe — CHANGES_REQUESTED** [review]

File: main/controller/thread_monitor.py [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+        ----------
+        threadlist : DICT
+            List of thread handles
+        Returns
+        -------
+        None.
+        '''
+        logging.info('Beginning thread monitoring')
+        while self.run_th_monitor:
+            for th_name in threadlist.keys():
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: you took out the threadlist argument (good), but now threadlist is undefined.
what I meant for you to do was make threadlist an instance attribute, so you need to rewrite your init method like so:

def \_\_init\_\_(self, th_dict):
    self.threadlist = th_dict
    .... [all the stuff you already have]
    
 then in the run method you can just call self.threadlist
 this also means that when you instantiate the monitor in observation run, you need to do self.monitor = Monitor(th_dict)

File: main/observing/observation_run.py [hunk 37]
```diff
@@ -84,6 +87,13 @@ def __init__(self, observation_request_list, image_directory, shutdown_toggle, c
         self.guider.start()
         self.gui.start()
 
+        self.th_dict = {'camera': self.camera, 'telescope': self.telescope,
+                        'dome': self.dome, 'focuser': self.focuser, 'flatlamp': self.flatlamp,
+                        'Conditions-Th': self.conditions, 'guider': self.guider,
+                        }
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py:96]: you are missing some threads here....there is also focus_procedures and GUI threads
also why do you use the official 'Conditions-Th' name for that one and none of the other ones?  all the threads are named similarly like 'Camera-Th', 'FlatLamp-Th', etc.

File: main/observing/observation_run.py [hunk 38]
```diff
@@ -229,12 +245,18 @@ def check_start_time(self, ticket):
         None.
         """
         current_time = datetime.datetime.now(self.tz)
+        time.sleep(10)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: why did you add a sleep time here?

File: main/controller/thread_monitor.py [hunk 39]
```diff
@@ -0,0 +1,47 @@
...
+                if not threadlist[th_name].is_alive():
+                    if not threadlist[th_name].name in self.crashed:
+                        self.crashed.append(threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(threadlist[th_name].name))
+                    self.threadcrash.set()
+
+
+    def pass_dict(self, th_dict):
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py]: this method is unneeded if you do what I explain above

File: main/observing/observation_run.py [hunk 40]
```diff
@@ -630,3 +662,62 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        Returns
+        -------
+        None
+        '''
+        threadlist = self.monitor.crashed
+        if self.monitor.threadcrash.isSet():
+            for thname in threadlist:
+                self.restart(thname)
+        else:
+            logging.info('All threads OK')
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: make this a debug statement

File: main/observing/observation_run.py [hunk 41]
```diff
@@ -630,3 +662,62 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        Description
+        -----------
+        Redefines and then restarts the inputted thread
+        
+        Parmeters
+        --------
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+        logging.info('Restarting thread {}'.format(thname))
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: this should probably be a logging.error

File: resources/weather_status/Radar coordinates.txt [hunk 42]
```diff
@@ -0,0 +1,35 @@
+Bottom-Right: xyz=(292,392,10)
```
Comment (Michael-Reefe) [inline resources/weather_status/Radar coordinates.txt]: don't include this in the commit

File: resources/weather_status/radar.txt [hunk 43]
```diff
@@ -0,0 +1,234 @@
+<!doctype html>
```
Comment (Michael-Reefe) [inline resources/weather_status/radar.txt]: don't include this in the commit either

File: resources/weather_status/weather.txt [hunk 44]
```diff
@@ -0,0 +1 @@
...t color=#ff4500>----</font></p></small></font></strong></font></td>\r\n\r\n<tr>\r\n <td  bordercolor="#000000" style="border: thin solid rgb(0,0,0)"  valign=middle><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#228b22><small><p>Low Wind Chill</p></small></font></strong></font></td>\r\n <td align=left bordercolor="#000000" style="border: thin solid rgb(0,0,0)" ><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#ff4500><small><p>23.0\xb0F<font color=#228b22>&nbsp;\r\n&nbsp;at&nbsp;&nbsp;</font><font color=#ff4500> 3:45a</font></p> </small></font></strong></font></td>\r\n </tr>\r\n<tr>\r\n <td  bordercolor="#000000" style="border: thin solid rgb(0,0,0)"  valign=middle><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#228b22><small><p>High Heat Index</p></small></font></strong></font></td>\r\n <td align=left bordercolor="#000000" style="border: thin solid rgb(0,0,0)" ><font face="verdana,\r\n Arial, Helvetica"><strong><font color=#ff4500><small><p>38.0\xb0F<font color=#228b22>&nbsp;\r\n&nbsp;at&nbsp;&nbsp;</font><font color=#ff4500> 1:01p</font></p> </small></font></strong></font></td>\r\n </tr>\r\n\r\n </table>\r\n\r\n</BODY>\r\n</HTML>\r\n'
```
Comment (Michael-Reefe) [inline resources/weather_status/weather.txt]: don't include in commit

**Review by Michael-Reefe — CHANGES_REQUESTED** [review]
A lot of my comments from the last review still apply (mostly removing unnecessary self.threadcheck() calls).  Please make sure to look through all of the comments carefully.  [Go to "Files Changed" to see the file structure and exactly where I've commented on each file]

File: main/observing/observation_run.py (hunk 23 above)
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: you correctly added the th_dict as an init parameter for monitor in th_monitor.py, but now you need to actually pass it in when you instantiate it, so you need to define self.th_dict _before_ this and then do self.monitor = Monitor(self.th_dict)

File: main/controller/thread_monitor.py [hunk 45]
```diff
@@ -0,0 +1,43 @@
...
+        -------
+        None.
+        '''
+        logging.debug('Beginning thread monitoring')
+        while self.run_th_monitor:
+            for th_name in self.threadlist.keys():
+                if not self.threadlist[th_name].is_alive():
+                    if not self.threadlist[th_name].name in self.crashed:
+                        self.crashed.append(self.threadlist[th_name].name)
+                        logging.error('{} thread has raised an exception'.format(self.threadlist[th_name].name))
```
Comment (Michael-Reefe) [inline main/controller/thread_monitor.py:39]: you still need to add a time.sleep here

File: main/observing/observation_run.py [hunk 46]
```diff
@@ -335,6 +352,9 @@ def focus_target(self, ticket):
             focus_exposure = 30
         self.focus_procedures.onThread(self.focus_procedures.startup_focus_procedure, focus_exposure,
                                        self.filterwheel_dict[focus_filter], self.image_directories[ticket])
+        while not self.focus_procedures.focus_isfinished.isSet():
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: we already have an event for this, self.focus_procedures.focused
just use that one

File: main/controller/focuser_procedures.py [hunk 47]
```diff
@@ -55,6 +55,7 @@ def __init__(self, focus_obj, camera_obj, conditions_obj):
        
         self.focused = threading.Event()
         self.continuous_focusing = threading.Event()
+        self.focus_isfinished = threading.Event()
```
Comment (Michael-Reefe) [inline main/controller/focuser_procedures.py]: self.focused is already the same thing, just use that one
(plus you never have anywhere where you 'set' this event so it wouldn't work anyways)

File: main/observing/observation_run.py [hunk 48]
```diff
@@ -235,6 +249,7 @@ def check_start_time(self, ticket):
             current_epoch_milli = time_utils.datetime_to_epoch_milli_converter(current_time)
             start_time_epoch_milli = time_utils.datetime_to_epoch_milli_converter(ticket.start_time)
             time.sleep((start_time_epoch_milli - current_epoch_milli) / 1000)
+        time.sleep(5)
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: what is this extra sleep for?

File: main/observing/observation_run.py [hunk 49]
```diff
@@ -301,7 +318,7 @@ def observe(self):
             (taken, total) = self.run_ticket(ticket)
             logging.info("{} out of {} exposures were taken for {}.  Moving on to next target.".format(taken, total,
                                                                                                        ticket.name))
-
+            self.focus_procedures.focus_isfinished.clear()
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: see other comments, you don't need this extra event

File: main/observing/observation_run.py [hunk 50]
```diff
@@ -614,6 +637,7 @@ def _shutdown_procedure(self, calibration, cooler=True):
         """
         logging.info("Shutting down observatory.")
         time.sleep(5)
+        self.monitor.run_th_monitor = False
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: Do NOT add this here.  _shutdown_procedure is called during observations when we temporarily close up due to weather, and you don't want to stop the monitor at those times.  Place this at the end of stop_threads() which is only called at the very very end.

File: main/observing/observation_run.py [hunk 51]
```diff
@@ -630,3 +654,70 @@ def _shutdown_procedure(self, calibration, cooler=True):
...
+        Description
+        -----------
+        Redefines and then restarts the inputted thread
+        
+        Parmeters
+        --------
+        thname : Handle
+            Handle of the original thread to restart
+        '''
+        logging.error('Restarting thread {}'.format(thname))
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py:709]: this method looks good now

File: main/observing/observation_run.py [hunk 52]
```diff
@@ -140,6 +152,7 @@ def everything_ok(self):
 
             else:
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py:160]: don't forget to add a threadcheck under the "while current_time < sunset_time" loop (currently line 146)

File: main/observing/observation_run.py (hunk 29 above)
Comment (Michael-Reefe) [reply main/observing/observation_run.py]: this one is still unnecessary

**Review by Michael-Reefe — CHANGES_REQUESTED** [review]
Still not quite ready, see comments

File: main/observing/observation_run.py [hunk 53]
```diff
@@ -630,3 +652,70 @@ def _shutdown_procedure(self, calibration, cooler=True):
             self.take_calibration_images()
         if cooler:
             self.camera.onThread(self.camera.cooler_set, False)
+        logging.info('Stopping thread monitoring')
+        self.monitor.run_th_monitor = False
+        logging.debug('Number of thread restarts: {}'.format(self.monitor.n_restarts))
```
Comment (Michael-Reefe) [inline main/observing/observation_run.py]: I saw you added this to the stop_threads method but you still need to get rid of this

File: main/controller/thread_monitor.py (hunk 45 above)
Comment (Michael-Reefe) [reply main/controller/thread_monitor.py:39]: still need this

File: main/observing/observation_run.py (hunk 23 above)
Comment (Michael-Reefe) [reply main/observing/observation_run.py]: don't forget about this

**Review by Michael-Reefe — CHANGES_REQUESTED** [review]
